### ISSUE TITLE
tf.keras: regularizer does not add to total losses

### ISSUE THREAD

**Issue by scotthuang1989** [tensorflow/tensorflow#21587]
I try to add l2 regularizer for tf.keras.layers.CuDNNLSTM,  here is my code for creating layers:

```
   regulizer = tf.keras.regularizers.l2(l=1) 
    net = tf.keras.layers.CuDNNLSTM(lstm_units,
                                    kernel_initializer=tf.keras.initializers.glorot_normal(),
                                    unit_forget_bias=True,
                                    recurrent_initializer=tf.keras.initializers.glorot_normal(),
                                    kernel_regularizer=regulizer,
                                    return_sequences=True)(input_data)
```
after I create my graph, I gather all the losses with 
`tf.get_collection(tf.GraphKeys.LOSSES)`

but 
**1. only  mse error is in the return list.
2. the initial losses is same as without regularizer even if I set regularizer co-efficient very big.**

Does someone know how to apply regularizer to keras layer in tensorflow?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by scotthuang1989** [comment]
I think the only relevant information is tensorflow version:

tensorflow: 1.9.0

**Comment by scotthuang1989** [comment]
@fchollet 

I check source code of tf.keras,  if I add a regularizer to a layer, the only thing tf.keras do is add regularizer loss to  **self._losses** of **Layer** class:

```
tensorflow/python/keras/engine/base_layer.py:

class Layer(checkpointable.CheckpointableBase):
      .....................
      def add_loss(self, losses, inputs=None):
           .....................
           self._losses += losses
```

if I train this graph with keras API, keras will collect all losses when I compile model. But this will not work if I train with native tensorflow API.

In the native tensorflow implementation, tensorflow will add regularizer loss to this collection:

**REGULARIZATION_LOSSES**

```
tensorflow\python\layers\base.py:

  def add_loss(self, losses, inputs=None):
    previous_losses_length = len(self._losses)
    super(Layer, self).add_loss(losses, inputs=inputs)
    # TODO(fchollet): deprecate collection below.
    new_losses = self._losses[previous_losses_length:]
    _add_elements_to_collection(new_losses, ops.GraphKeys.REGULARIZATION_LOSSES)
```

Is my understanding correct, or I did it in a wrong way? If this is a bug I  think it is very easy to fix.

**Comment by KuribohG** [comment]
Hello!
I have the same problem. Did you solve it? Currently I have not found this problem in 1.8.0, but it happens in 1.10.0.

**Comment by scotthuang1989** [comment]
I have some hot fix to solve this, just add regularization loss to total loss. but not sure if this is right (right way of doing it in tensorflow, It should be right in math).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jrabary** [comment]
@scotthuang1989 As far as I understand, Tensorflow will depreciate the use of global variables. I think the better way to handle this is then:
```
total_loss = loss + tf.add_n(model.losses)
```
Where the second term contains weights and activations regulirization losses

**Comment by jrabary** [comment]
@scotthuang1989 Most of the time you have several layers. tf.keras.Model helps to compose these layers. For example tf.keras.Sequential is a subclass of tf.keras.Model. So, I think you should really build your neural net using tf.keras.Model c.f  https://www.tensorflow.org/guide/keras#model_subclassing

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tranhungnghiep** [comment]
@ssubraveti In `tensorflow`, regularization loss is added to regularization loss key, not the main loss key. 
Unless you use high level API like keras Model or Estimator, all losses have to be collected explicitly. So your total loss would look like:

    loss = main_loss + tf.losses.get_regularization_loss()

@fchollet This is not a `tensorflow` issue, it should be closed.

**Comment by scotthuang1989** [comment]
@tranhungnghiep , if you check source code of tf.losses.get_regularization_loss.  you will find that this function also get regularization loss from **REGULARIZATION_LOSSES** collection

`ops.get_collection(ops.GraphKeys.REGULARIZATION_LOSSES, scope)`

As I mentioned in my comment above ,  keras layer don't add regularization loss to this collection. I am able to fix this, but I am not sure if this is the design or a bug?


> @fchollet
> 
> I check source code of tf.keras, if I add a regularizer to a layer, the only thing tf.keras do is add regularizer loss to **self._losses** of **Layer** class:
> 
> ```
> tensorflow/python/keras/engine/base_layer.py:
> 
> class Layer(checkpointable.CheckpointableBase):
>       .....................
>       def add_loss(self, losses, inputs=None):
>            .....................
>            self._losses += losses
> ```
> if I train this graph with keras API, keras will collect all losses when I compile model. But this will not work if I train with native tensorflow API.
> 
> In the native tensorflow implementation, tensorflow will add regularizer loss to this collection:
> 
> **REGULARIZATION_LOSSES**
> 
> ```
> tensorflow\python\layers\base.py:
> 
>   def add_loss(self, losses, inputs=None):
>     previous_losses_length = len(self._losses)
>     super(Layer, self).add_loss(losses, inputs=inputs)
>     # TODO(fchollet): deprecate collection below.
>     new_losses = self._losses[previous_losses_length:]
>     _add_elements_to_collection(new_losses, ops.GraphKeys.REGULARIZATION_LOSSES)
> ```
> Is my understanding correct, or I did it in a wrong way? If this is a bug I think it is very easy to fix.

**Comment by tranhungnghiep** [comment]
@scotthuang1989 I think you are right. tf's `add_loss()` adds regularization loss to `GraphKeys.REGULARIZATION_LOSSES`, but keras' `add_loss()` doesn't. So `tf.losses.get_regularization_loss()` works for tf layer but not keras layer. For keras layer, you should call `layer._losses` or `layer.get_losses_for()`.

I also see @fchollet's comment that `GraphKeys.REGULARIZATION_LOSSES` is deprecated, so the correct way is to go through each and every layer/variable and get loss out manually. You can also use tf instead of keras, it works, for now.

@fchollet There is really a mismatch in tf and keras behaviors. And regarding your comment on deprecation, please consider either update the code or the documentation to keep consistency and avoid confusion. Thanks.

**Comment by SayHiRay** [comment]
I just encountered the same issue. Apparently layers in native tf and keras are different in their ways of dealing with regularization loss.

I found an example of dealing with such regularization loss on tf official example: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/examples/densenet/densenet_graph_test.py

They use `tf.keras.model.losses` to collect such loss. In general I think it is nowadays a good practice to subclass tf.keras.model API for building tensorflow models. If you do have mixed sources of regularization losses, you may consider collect them using `tf.losses.get_regularization_loss` and `tf.keras.model.losses` respectively.

**Comment by scotthuang1989** [comment]
although no official answer, but it does have some workaround, I currently manually add regularizer to each layer. In future I try avoid mix tf.layer and tf.keras.layers.  I am gonna close this.

**Comment by isaacgerg** [comment]
I also found this issue today and opened a feature request.  Moreso, these attributes should be read-only unless the model state is updated when they are written to.

https://github.com/keras-team/keras/issues/12053
